exts/virtio_block/lib.rs

```diff
@@ -0,0 +1,313 @@
+#![no_std]
+
+extern crate alloc;
+#[macro_use]
+extern crate kerla_api;
+
+use core::mem::size_of;
+
+use alloc::boxed::Box;
+use kerla_api::address::{VAddr, PAddr};
+use kerla_api::driver::{DeviceProber, Driver, register_driver_prober, attach_irq};
+use kerla_api::driver::block::{BlockDriver, register_block_driver};
+use kerla_api::{info, warn};
+use kerla_api::mm::{alloc_pages, AllocPageFlags};
+use kerla_api::arch::PAGE_SIZE;
+
+use kerla_api::sync::SpinLock;
+use kerla_utils::alignment::align_up;
+
+use memoffset::offset_of;
+
+use alloc::sync::Arc;
+
+use virtio::device::{Virtio, VirtqDescBuffer, IsrStatus, VirtqUsedChain};
+
+use virtio::transports::{
+    virtio_mmio::VirtioMmio, virtio_pci_legacy::VirtioLegacyPci,
+    virtio_pci_modern::VirtioModernPci, VirtioAttachError, VirtioTransport,
+};
+
+const VIRTIO_BLK_F_SIZE: u64 = 1 << 6;
+const VIRTIO_REQUEST_QUEUE: u16 = 0;
+const MAX_BLK_SIZE: usize = 512;
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct VirtioBlockConfig {
+    capacity: u64,
+    size_max: u32,
+    seg_max: u32,
+    cylinders: u16,
+    heads: u8,
+    sectors: u8,
+    block_size: u32,
+    physical_block_exp: u8,
+    alignment_offset: u8,
+    min_io_size: u16,
+    opt_io_size: u32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct VirtioBlockRequest {
+    type_: u32,
+    reserved: u32,
+    sector: u64,
+}
+
+
+
+#[repr(u32)]
+enum RequestType {
+    Read = 0,
+    Write = 1
+}
+
+
+pub struct VirtioBlock {
+    virtio: Virtio,
+    request_buffer: VAddr,
+}
+
+impl VirtioBlock {
+    pub fn new(transport: Arc<dyn VirtioTransport>) -> Result<VirtioBlock, VirtioAttachError> {
+        let mut virtio = Virtio::new(transport);
+        virtio.initialize(VIRTIO_BLK_F_SIZE, 1)?;
+        // Read the block size
+        let block_size = virtio.read_device_config64(offset_of!(VirtioBlockConfig, capacity) as u16);
+        
+        info!("Block size is {} bytes", block_size);
+        
+        let ring_len = virtio.virtq(VIRTIO_REQUEST_QUEUE).num_descs() as usize;
+        
+        let buffer = alloc_pages(
+            (align_up(MAX_BLK_SIZE * ring_len, PAGE_SIZE)) / PAGE_SIZE,
+            AllocPageFlags::KERNEL 
+        ).unwrap().as_vaddr();
+
+        
+
+        Ok(VirtioBlock {
+            virtio: virtio,
+            request_buffer: buffer
+        
+        })
+
+
+    }
+
+    fn request_to_device(&mut self, request_type: RequestType, sector: u64, frame: &[u8]) {
+        let request_addr = self.request_buffer.add(MAX_BLK_SIZE);
+        let request_len = size_of::<VirtioBlockRequest>();
+        let read_addr = unsafe {PAddr::new(*frame.as_ptr() as usize)}; 
+
+        let read_len = frame.len();
+
+        let status_buffer = [0];
+        let status_addr =unsafe { PAddr::new(*status_buffer.as_ptr() as usize)};
```

Comment:
`status_buffer` will be freed when leaving this method while it may be modified later by the device (hypervisor). Allocate a memory buffer in advance just like `self.request_buffer`.